### ISSUE TITLE
are there plans to support interactive widgets? like ipywidgets?

### ISSUE THREAD

**Issue by cornhundred** [nteract/archived-desktop-app#1385]

**Comment by rgbkrk** [comment]
Thanks for posting. No one is currently working on supporting the current version of the ipython widgets. It's open for someone to work on.

When the version for JupyterLab is more finalized, I'll begin working on integration if no one else gets to it.

**Comment by cornhundred** [comment]
Thanks

**Comment by paulnakroshis** [comment]
Widgets are useful; in testing out NTERACT, I just discovered that they do not work. 
Also bqplot doesn't work in NTERACT; see https://nbviewer.jupyter.org/gist/anonymous/2923cbda7f825eb81a5fc50dc47a33e9
for example (last cell in notebook)

**Comment by rgbkrk** [comment]
If someone is able to support widgets, I'd be happy to have them.

**Comment by cogmind** [comment]
I also support the inclusion of ipywidgets. A "workaround" is to export to HTML:

    `ipywidgets.embed.embed_minimal_html`

By the way, I think it's great that `IPython.display.Markdown` is supported.

**Comment by spinicist** [comment]
I'm another one who would love to see widgets supported!

**Comment by jdfreder** [comment]
I'm looking into this now...

**Comment by captainsafia** [comment]
Commenting to remove stale tag. This is still something we need to do.

**Comment by rgbkrk** [comment]
Added the `pinned` tag so it sticks around.

**Comment by wadethestealth** [comment]
[`@jupyterlab/widgets`](https://github.com/jasongrout/jupyterlab/tree/master/src/widgets) that @rgbkrk mentioned during the meeting. Plus [`ipywidgets` docs](https://ipywidgets.readthedocs.io/en/stable/examples/Widget%20Low%20Level.html) that explain how they work under the hood. Also link to nteract/hydrogen#27

**Comment by jasongrout** [comment]
The jlab widget manager is at https://github.com/jupyter-widgets/ipywidgets/tree/master/packages/jupyterlab-manager

In addition to the docs, see the controls schema specification at https://github.com/jupyter-widgets/ipywidgets/blob/master/packages/schema/jupyterwidgetmodels.latest.md and the message spec at https://github.com/jupyter-widgets/ipywidgets/blob/master/packages/schema/messages.md#widget-messaging-protocol-version-2

**Comment by jasongrout** [comment]
Also, I'd point to the discussions with @williamstein when he implemented widgets for cocalc: https://github.com/jupyter-widgets/ipywidgets/issues?q=is%3Aissue+author%3Awilliamstein+is%3Aclosed

**Comment by motin** [comment]
TLDR of this thread so far:
- Interactive widgets that support Jupyter Lab will eventually be supported by nteract
- @jdfreder did some initial work related to widget cross messaging in December, which was great but ultimately did not end up with any working proof of concept of a jupyter widget in nteract
- Nteract maintainers had a meeting about this about 10 days ago, which shed some light on possible ways to implement this

Have I understood it correctly?

**Comment by motin** [comment]
> I also support the inclusion of ipywidgets. A "workaround" is to export to HTML:
> 
> `ipywidgets.embed.embed_minimal_html`

To clarify this workaround: I tried this for a couple of ipywidgets (qgrid, ipysheet, pivottablejs) and these were the results:

## [Qgrid](https://github.com/quantopian/qgrid)
```
import pandas as pd
import qgrid
from ipywidgets.embed import embed_minimal_html
from IPython.display import IFrame

df = pd.read_csv(
  "https://gist.githubusercontent.com/rgbkrk/a7984a8788a73e2afb8fd4b89c8ec6de/raw/db8d1db9f878ed448c3cac3eb3c9c0dc5e80891e/2015.csv"
)
widget = qgrid.show_grid(df)
embed_minimal_html('qgrid.html', views=[widget], title='QGrid')
IFrame("qgrid.html", width=1000, height=400)
```
Results: The datatable renders but all interactivity is broken. [QGrid 1.0 does not support static export](https://github.com/quantopian/qgrid/issues/110#issuecomment-336929573) since sorting and filtering is done server-side. Also crucially, there is no way to communicate with the rendered datatable and thus no way of getting information about the edited data (usually, this is achieved by running `widget.get_changed_df()`.

## [Ipysheet]()
```
import pandas as pd
from ipysheet import from_dataframe
from IPython.display import IFrame
from ipywidgets.embed import embed_minimal_html

df = pd.read_csv(
  "https://gist.githubusercontent.com/rgbkrk/a7984a8788a73e2afb8fd4b89c8ec6de/raw/db8d1db9f878ed448c3cac3eb3c9c0dc5e80891e/2015.csv"
)
sheet = from_dataframe(df)
view = sheet
embed_minimal_html('ipysheet-export.html', views=[view], title='ipysheet export')
IFrame("ipysheet-export.html", width=1000, height=400)
```
Results: The datatable renders but editing any cell results in a 1-2sec UI freeze. Again, there is no way to communicate with the rendered datatable and thus no way of getting information about the edited data. 

## [Jupyter Pivottable.js](https://github.com/nicolaskruchten/jupyter_pivottablejs)
```
from pivottablejs import pivot_ui
from IPython.display import HTML

df = pd.read_csv(
  "https://gist.githubusercontent.com/rgbkrk/a7984a8788a73e2afb8fd4b89c8ec6de/raw/db8d1db9f878ed448c3cac3eb3c9c0dc5e80891e/2015.csv"
)
pivot_ui(df,outfile_path="pivottablejs.html")
HTML("pivottablejs.html")
```
Results: This widget renders as a standalone interactive widget. The widget does not support communicating any changes back to the notebook even in ordinary Jupyter environments, so there is no change in functionality here.

**Comment by motin** [comment]
@rgbkrk @jasongrout @wadethestealth, what is the outline of what would be necessary to support interactive widgets? Some code pointers?

**Comment by wadethestealth** [comment]
@motin the first problem is CSP doesn't allow script injections, so we have to make sure we use local scripts to bypass that. Then we need to be able to create the html of the component to be rendered on the dom and link it to the widget state and behavior in the kernel so that when the kernel widget changes the Dom component changes too. I am pretty sure there is also a vice versa for Dom value change to kernel, but I don't use ipywidgets so I am unsure. The jupyter lab widgets linked above are a very good example implementation of what we will need. Ideally we would use react though.

**Comment by motin** [comment]
Thanks @wadethestealth for your reply. How does this relate to @jdfreder's [PR](https://github.com/nteract/nteract/pull/3642#issue-229971989), which seem to presume that the widget will be rendered in cross domain iframes?
CSP should not be an issue in Electron, or at least I would assume that it is possible to workaround it.

I started looking at the nteract source code and so far I get the following:
1. Different mimetypes are rendered by different transforms
2. The jupyter-widgets transform receives a `data` property, which can contain information like:
```
{"data":{"version_major":2,"version_minor":0,"model_id":"58d115637efc47ba8fbc9f81f0f77014"},"metadata":{},"theme":"light"}
```
Somehow, this data JSON should be rendered as a widget, using a similar approach as jupyterlab's widgetmanager :)

I am happy to take a look at this, but I am new to nteract development.

**Comment by wadethestealth** [comment]
That pr is some early work of trying to provide support. CSP is an issue in this cause because the implementation should also attempt to support nteract/hydrogen which is an atom package. Atom has CSP and I also think nteract added one too but I am unsure. Statement 1 is true. Statement 2 is kinda true. The data is the original message sent so that you can setup the component and the communication system between the Dom and the kernel widget (since they are separate objects and must stay in sync). While this should work in nteract, if you can get it normal browser Js and Dom to work we should be able to help it integrate into react in the nteract format.

**Comment by wadethestealth** [comment]
You should read some of the links posted here as they give big hints and insights into how this should be implemented

**Comment by motin** [comment]
Aha that explains the CSP restriction and the general outline, thanks again. I read the linked material a bit more thoroughly this time and there are plenty of good reference implementations. Tomorrow I’ll try to figure out the nteract format, ie how transforms gets set up in the nteract React app, how messaging between the kernels and React app work, and what the model_id parameter is. Baby steps... any pointers welcome :)

**Comment by captainsafia** [comment]
@motin The cross-domain iframes is an implementation strategy that targets usage of nteract in web environments where we want to avoid the widgets having access to things like cookies in the domain.

I spoke with @jdfreder recently and while that was the direction he was headed, it isn't complete necessary for our implementation. Personally, I would start with an approach that is sans iframe and then build the iframe in as an isolation boundary.

With respect to how transforms get set up, they are React packages that are imported and configured as renderers in the `transforms` property of the state.

The messaging is handled by `enchannel` which exposes an RxJS Observable that can be subscribed to for sending events to the kernel.

**Comment by captainsafia** [comment]
To those following this thread, below is an RFC for ipywidgets support.

https://github.com/nteract/nteract/issues/4573

**Comment by captainsafia** [comment]
Closing this because we've added support for the canonical set of ipywidgets to nteract. It'll be shipped in the next desktop release.
